File: papersrc/group-productivity.R

# load libraries

library(lattice)

# functions

## setting the data to plot

plot.prod <- function(m.raw, m.cart, measure, ylab, do.plot=TRUE) {
	pd = list()
	for(n in sort(unique(m.cart$groupsize))) {
		m <- sub("prod", "", measure)
		m <- sub("s$", "", m)
		pd[[paste("r", n, sep="")]] = m.raw[m.raw$measure == m,
																				paste("X", n, sep="")]
		pd[[paste("c", n, sep="")]] = m.cart[m.cart$groupsize == n, measure]
	}
	if(do.plot) {
		l <- unlist(pd)
		s <- sub("^([rc])([1-8]0*).*", "\\1\\2", names(l))

		a <- 1:(1.5*length(pd))
		a <- a[a %% 3 != 0]

		plot(1,1, xlim=range(a), ylim=range(l), log="y", type="n", xlab="group size",
				 ylab=ylab, xaxt="n")
		axis(side=1, at=rowMeans(matrix(a, ncol=2, byrow=TRUE)),
				 labels=sort(unique(m.cart$groupsize)))
		for(i in 1:length(a)) {
			if(a[i] %% 3 == 1) {
				#mcol <- rgb(0.5, 0, 0, 0.05)
				mcol <- rgb(1, 0.6, 0.2, 0.05)
				#bcol <- rgb(1, 0.6, 0.2, 1)
				bcol <- "black"
			} else {
				mcol <- rgb(0, 0.5, 0, 0.25)
				bcol <- "black"
			}
			points(jitter(rep(a[i], length(pd[[i]])), amount=0.25), pd[[i]], pch=16, 
						 col=mcol)
			boxplot(pd[[i]], outline=FALSE, log="y", add=TRUE, at=a[i], col=mcol,
							border=bcol, yaxt="n")
		}
	}
	return(invisible(pd))
}

cartel.z.scores <- function(p) {
	l <- length(p)
	ns <- as.numeric(unique(sub("^[rc]", "", names(p))))
	m.z <- list()
	n.c <- list()
	for(n in ns) {
		mr <- p[[paste("r", n, sep="")]]
		mc <- p[[paste("c", n, sep="")]]
		tot <- c(mr, mc)
		gr <- rep(c("r", "c"), c(length(mr), length(mc)))
		stot <- scale(tot)
		m.z[[paste("c", n, sep="")]] <- stot[gr=="c"]
		n.c[[paste("c", n, sep="")]] <- rep(n, length(mc))
	}
	unlist(m.z)
}

cartel.z.diff <- function(p) {
	l <- length(p)
	ns <- as.numeric(unique(sub("^[rc]", "", names(p))))
	m.z <- numeric(length(ns))
	n.c <- numeric(length(ns))
	i <- 1
	for(n in ns) {
		mr <- p[[paste("r", n, sep="")]]
		mc <- p[[paste("c", n, sep="")]]
		tot <- c(mr, mc)
		gr <- rep(c("r", "c"), c(length(mr), length(mc)))
		stot <- scale(tot)
		m.z[i] <- mean(stot[gr=="c"])
		n.c[i] <- length(mc)
		i <- i+1
	}
	weighted.mean(m.z, n.c)
}

cartel.diff <- function(p, Q=0.5) {
	l <- length(p)
	ns <- as.numeric(unique(sub("^[rc]", "", names(p))))
	m.z <- numeric(length(ns))
	n.c <- numeric(length(ns))
	i <- 1
	for(n in ns) {
		mr <- p[[paste("r", n, sep="")]]
		mc <- p[[paste("c", n, sep="")]]
		m.z[i] <- sum(mc < quantile(mr, Q))
		n.c[i] <- length(mc)
		i <- i+1
	}
	sum(m.z)/sum(n.c)
}

cartel.median.diff <- function(p) {
	l <- length(p)
	ns <- as.numeric(unique(sub("^[rc]", "", names(p))))
	m.z <- list()
	n.c <- numeric(length(ns))
	i <- 1
	for(n in ns) {
		mr <- p[[paste("r", n, sep="")]]
		mc <- p[[paste("c", n, sep="")]]
		m.z[[paste("c", n, sep="")]] <- (mc-median(mr))/median(mr)
		n.c[i] <- length(mc)
		i <- i+1
	}
	median(unlist(m.z))
}

cartel.m.scores <- function(p) {
	l <- length(p)
	ns <- as.numeric(unique(sub("^[rc]", "", names(p))))
	m.z <- list()
	for(n in ns) {
		mr <- p[[paste("r", n, sep="")]]
		mc <- p[[paste("c", n, sep="")]]
		tot <- c(mr, mc)
		gr <- rep(c("r", "c"), c(length(mr), length(mc)))
		stot <- (tot-median(tot))/IQR(tot)
		m.z[[paste("c", n, sep="")]] <- stot[gr=="c"]
	}
	unlist(m.z)
}

compare.rnd2cart <- function(p, Qs=c(0,0.05, 0.25, 0.5, 0.75, 0.95, 1)) {
	d <- numeric(length(Qs))
	i <- 1
	for(Q in Qs) {
		d[i] <- cartel.diff(p, Q)
		i <- i+1
	}
	data.frame(Q=Qs, d=d)
}

opar <- par()

# MTMT

## read data

m.raw <- read.csv("../analyse-publications/MTMT/MTMTpubmat-productivity_raw.csv")
m.cart <- read.csv("../analyse-publications/MTMT/MTMTpubmat-cartel_productivity.csv")

## process and plot data


pdf(file="../paperfigs/group_productivity.pdf", width=7, height=10)
layout(matrix(1:8, ncol=2))
par(mar=c(4,4,0,1)+0.1)

prods <- list()
prods[["groupprod"]] <- plot.prod(m.raw, m.cart, "groupprod",
																	"group productivity")
prods[["npapers"]] <- plot.prod(m.raw, m.cart, "npapers",
																"mean number of papers")
prods[["wpapers"]] <- plot.prod(m.raw, m.cart, "wpapers",
																"mean weighted number of papers")

gp <- compare.rnd2cart(prods$groupprod, seq(0,1,0.1))
np <- compare.rnd2cart(prods$npapers, seq(0,1,0.1))
wp <- compare.rnd2cart(prods$wpapers, seq(0,1,0.1))

## plot data

plot(gp, xlim=c(0,1), ylim=c(0,1), type="o", pch=16,
		 xlab="quantiles of random samples",
		 ylab="proportion in subspicious groups")
grid()
abline(a=0, b=1, lty=2)
lines(np, type="o", pch=17, col=2)
lines(wp, type="o", pch=18, col=3)
legend("bottomright", legend=c("group productivity", "number of papers",
															 "weighted number of papers"), col=1:3,
			 pch=16:18, lty=1, bty="n", title="measures")

# dblp

## read data

m.raw <- read.csv("../analyse-publications/dblp/dblppubmat-productivity_raw.csv")
m.cart <- read.csv("../analyse-publications/dblp/dblppubmat-cartel_productivity.csv")

## process and plot data

prods <- list()
prods[["groupprod"]] <- plot.prod(m.raw, m.cart, "groupprod",
																	"group productivity")
prods[["npapers"]] <- plot.prod(m.raw, m.cart, "npapers",
																"mean number of papers")
prods[["wpapers"]] <- plot.prod(m.raw, m.cart, "wpapers",
																"mean weighted number of papers")

gp <- compare.rnd2cart(prods$groupprod, seq(0,1,0.1))
np <- compare.rnd2cart(prods$npapers, seq(0,1,0.1))
wp <- compare.rnd2cart(prods$wpapers, seq(0,1,0.1))

# plot data

plot(gp, xlim=c(0,1), ylim=c(0,1), type="o", pch=16,
		 xlab="quantiles of random samples",
		 ylab="proportion in subspicious groups")
grid()
abline(a=0, b=1, lty=2)
lines(np, type="o", pch=17, col=2)
lines(wp, type="o", pch=18, col=3)
legend("bottomright", legend=c("group productivity", "number of papers",
															 "weighted number of papers"), col=1:3,
			 pch=16:18, lty=1, bty="n", title="measures")

layout(1)
par(opar)
dev.off()

#z.gp = cartel.z.scores(prods$groupprod)
#z.np = cartel.z.scores(prods$npapers)
#z.wp = cartel.z.scores(prods$wpapers)
#
#boxplot(z.gp, z.np, z.wp, outline=FALSE)
#
#m.gp = cartel.m.scores(prods$groupprod)
#m.np = cartel.m.scores(prods$npapers)
#m.wp = cartel.m.scores(prods$wpapers)
#
#boxplot(m.gp, m.np, m.wp, outline=FALSE)

File: papersrc/group-productivity.jl

using HypothesisTests, GLM

include("../analyse-publications/PaperPump.jl")
include("../analyse-publications/CommunitiesNetworks.jl")

matplotlib.use("PDF")

function plprod(sbplt, cartdata, variable, rnddata, xlab="")
	cad = cartdata[:car][variable]
	maxx = (mx1 = maximum(cad)) < (mx2 = maximum(rnddata)) ? mx2 : mx1
	meanx = round(median(cad), digits=1)
	sbplt.hist(cad, density=true, label="tight groups, M = $meanx",
							alpha=0.25, bins=0:5:maxx, log=false)
	meanx = round(median(rnddata), digits=1)
	sbplt.hist(rnddata, density=true, label="random groups, M = $meanx",
							alpha=0.25, bins=0:5:maxx, log=false)
	sbplt.legend(fontsize=:small)
	sbplt.set_xlabel(xlab)
	sbplt.set_ylabel("density")
	ins = sbplt.inset_axes([0.65, 0.05, 0.3, 0.65])
	ins.boxplot(cad, positions=[1], showfliers=false, #labels=(""),
							patch_artist=true, boxprops = Dict("facecolor" => "C0",
																								 "alpha" => 0.25))
	ins.boxplot(rnddata, positions=[2], showfliers=false, #labels=(""),
							patch_artist=true, boxprops = Dict("facecolor" => "C1",
																								 "alpha" => 0.25))
	ins.set_xticklabels(labels="")
	ins.tick_params(axis="x", bottom=false)
	ins.tick_params(axis="y", labelsize=:small)
end


function toppers(cartdata, variable, rnddata, Q)
	cad = cartdata[:car][variable]
	return sum(cad .> quantile(rnddata, Q)) / length(cad)
end

function createDF(proddata)
	proddata[:car][:,:grtype] .= "tightgroup"
	nc = ncol(proddata[:raw])
	i = proddata[:raw][:measure] .== "group"
	rgrp = flat(Matrix(proddata[:raw][i, 1:(nc-1)]))
	i = proddata[:raw][:measure] .== "npaper"
	rnpa = flat(Matrix(proddata[:raw][i, 1:(nc-1)]))
	i = proddata[:raw][:measure] .== "wpaper"
	rwpa = flat(Matrix(proddata[:raw][i, 1:(nc-1)]))
	rgrsize = names(proddata[:raw])[1:(nc-1)]
	rgrsize = parse.(Int, rgrsize)
	rgrsize = flat(map(x -> repeat([x], Int(nrow(proddata[:raw])/3)), rgrsize))
	rdf = DataFrame(:groupsize => rgrsize, :groupprod => rgrp, :npapers => rnpa,
									:wpapers => rwpa)
	rdf[:,:grtype] = "random"
	resdf = vcat(proddata[:car], rdf)
	return(resdf)
end

function calcquantile(df::DataFrame, measure::Symbol)
	grtype = df.grtype .== "tightgroup"
	tgdf = df[grtype,:]
	rgdf = df[.!grtype,:]
	GS = unique(rgdf.groupsize)
	mD = Dict{Int, Array{Float64, 1}}()
	for n in GS
		mD[n] = rgdf[rgdf.groupsize .== n, measure]
	end
	Qs = zeros(nrow(tgdf))
	for i in 1:nrow(tgdf)
		m = tgdf[i, measure]
		Qs[i] = sum(m .>= mD[tgdf.groupsize[i]]) ./ length(mD[tgdf.groupsize[i]])
	end
	return Qs
end

pMTMT = Dict()
pMTMT[:raw] = CSV.read("../analyse-publications/MTMT/MTMTpubmat-productivity_raw.csv",
											 DataFrame)
pMTMT[:car] = CSV.read("../analyse-publications/MTMT/MTMTpubmat-cartel_productivity.csv",
											 DataFrame)

pdblp = Dict()
pdblp[:raw] = CSV.read("../analyse-publications/dblp/dblppubmat-productivity_raw.csv",
											 DataFrame)
pdblp[:car] = CSV.read("../analyse-publications/dblp/dblppubmat-cartel_productivity.csv",
											 DataFrame)

m = Matrix(pMTMT[:raw][pMTMT[:raw][:measure] .== "group",
											 1:(size(pMTMT[:raw], 2)-1)])
gMTMT = reshape(m, prod(size(m)))

m = Matrix(pMTMT[:raw][pMTMT[:raw][:measure] .== "npaper",
											 1:(size(pMTMT[:raw], 2)-1)])
nMTMT = reshape(m, prod(size(m)))

m = Matrix(pMTMT[:raw][pMTMT[:raw][:measure] .== "wpaper",
											 1:(size(pMTMT[:raw], 2)-1)])
wMTMT = reshape(m, prod(size(m)))

m = Matrix(pdblp[:raw][pdblp[:raw][:measure] .== "group",
											 1:(size(pdblp[:raw], 2)-1)])
gdblp = reshape(m, prod(size(m)))

m = Matrix(pdblp[:raw][pdblp[:raw][:measure] .== "npaper",
											 1:(size(pdblp[:raw], 2)-1)])
ndblp = reshape(m, prod(size(m)))

m = Matrix(pdblp[:raw][pdblp[:raw][:measure] .== "wpaper",
											 1:(size(pdblp[:raw], 2)-1)])
wdblp = reshape(m, prod(size(m)))

# number of authors in tight groups
sum(pMTMT[:car].groupsize)
sum(pdblp[:car].groupsize)

# number of tight groups
nrow(pMTMT[:car])
nrow(pdblp[:car])

sbplts = PyPlot.subplots(3,2,figsize=(7,7))
plprod(sbplts[2][1], pMTMT, :npapers, nMTMT, "mean number of papers")
plprod(sbplts[2][2], pMTMT, :groupprod, gMTMT, "mean group productivity")
plprod(sbplts[2][3], pMTMT, :wpapers, wMTMT, "mean weighted number of papers")
plprod(sbplts[2][4], pdblp, :npapers, ndblp, "mean number of papers")
plprod(sbplts[2][5], pdblp, :groupprod, gdblp, "mean group productivity")
plprod(sbplts[2][6], pdblp, :wpapers, wdblp, "mean weighted number of papers")
tight_layout()

savefig("../paperfigs/group_productivity.pdf")

dfMTMT = createDF(pMTMT)
dfdblp = createDF(pdblp)

#dfMTMT[:,:s_groupprod] = (dfMTMT.groupprod .- mean(dfMTMT.groupprod)) ./ std(dfMTMT.groupprod)
#dfdblp[:,:s_groupprod] = (dfdblp.groupprod .- mean(dfdblp.groupprod)) ./ std(dfdblp.groupprod)

dfMTMT[:,:s_groupprod] = (dfMTMT.groupprod .- mean(dfMTMT.groupprod))
dfdblp[:,:s_groupprod] = (dfdblp.groupprod .- mean(dfdblp.groupprod))

fmMTMT0 = lm(@formula(npapers ~ grtype * groupprod), dfMTMT)
fmMTMT1 = lm(@formula(wpapers ~ grtype * groupprod), dfMTMT)
fmdblp0 = lm(@formula(npapers ~ grtype * groupprod), dfdblp)
fmdblp1 = lm(@formula(wpapers ~ grtype * groupprod), dfdblp)

i = dfMTMT.groupsize .== 2
fmMTMT20 = lm(@formula(npapers ~ grtype * s_groupprod), dfMTMT[i, :])
fmMTMT21 = lm(@formula(wpapers ~ grtype * s_groupprod), dfMTMT[i, :])
i = dfdblp.groupsize .== 2
fmdblp20 = lm(@formula(npapers ~ grtype * s_groupprod), dfdblp[i, :])
fmdblp21 = lm(@formula(wpapers ~ grtype * s_groupprod), dfdblp[i, :])

figure(figsize=(12,5))
subplot(1,2,1)
newX = DataFrame(:groupprod =>
								 repeat([minimum(dfMTMT.groupprod), maximum(dfMTMT.groupprod)],
												2))
newX[:,:grtype] = vcat(repeat(["tightgroup"], Int(nrow(newX)/2)),
											 repeat(["random"], Int(nrow(newX)/2)))
#close("all")
i = dfMTMT.grtype .== "random"
PyPlot.scatter(dfMTMT.groupprod[i], dfMTMT.npapers[i], alpha=0.1,
							 label="random groups")
i = dfMTMT.grtype .== "tightgroup"
PyPlot.scatter(dfMTMT.groupprod[i], dfMTMT.npapers[i], alpha=0.1,
							 label="tight groups")
i = newX.grtype .== "random"
PyPlot.plot(newX.groupprod[i], predict(fmMTMT0, newX[i,:]),
							 label="random groups, estimates")
i = newX.grtype .== "tightgroup"
PyPlot.plot(newX.groupprod[i], predict(fmMTMT0, newX[i,:]),
						label="tight groups, estimates")
title("MTMT")
legend()
xlabel("average group productivity")
ylabel("average number of papers")
tight_layout()
#savefig("../work/screen.pdf")
subplot(1,2,2)
newX = DataFrame(:groupprod =>
								 repeat([minimum(dfdblp.groupprod), maximum(dfdblp.groupprod)],
												2))
newX[:,:grtype] = vcat(repeat(["tightgroup"], Int(nrow(newX)/2)),
											 repeat(["random"], Int(nrow(newX)/2)))
#close("all")
i = dfdblp.grtype .== "random"
PyPlot.scatter(dfdblp.groupprod[i], dfdblp.npapers[i], alpha=0.1,
							 label="random groups")
i = dfdblp.grtype .== "tightgroup"
PyPlot.scatter(dfdblp.groupprod[i], dfdblp.npapers[i], alpha=0.1,
							 label="tight groups")
i = newX.grtype .== "random"
PyPlot.plot(newX.groupprod[i], predict(fmdblp0, newX[i,:]),
							 label="random groups, estimates")
i = newX.grtype .== "tightgroup"
PyPlot.plot(newX.groupprod[i], predict(fmdblp0, newX[i,:]),
						label="tight groups, estimates")
title("dblp")
legend()
xlabel("average group productivity")
ylabel("average number of papers")
tight_layout()
savefig("../paperfigs/paper_sharing-relations.pdf")

# estimating c

i = dfMTMT.grtype .== "tightgroup"
#i = repeat([true], nrow(dfMTMT))
alfa = dfMTMT.npapers[i] ./ dfMTMT.groupprod[i]
cMTMT = (alfa .- 1) ./ (dfMTMT.groupsize[i] .- 1)
i = dfdblp.grtype .== "tightgroup"
#i = repeat([true], nrow(dfdblp))
alfa = dfdblp.npapers[i] ./ dfdblp.groupprod[i]
cdblp = (alfa .- 1) ./ (dfdblp.groupsize[i] .- 1)

describe(cMTMT)
describe(cdblp)

figure()
PyPlot.hist(cMTMT, alpha=0.25, density=true, bins=0.2:0.05:1, label="MTMT")
PyPlot.hist(cdblp, alpha=0.25, density=true, bins=0.2:0.05:1, label="dblp")
legend()
xlabel("level of authorship sharing")
ylabel("density")
tight_layout()
savefig("../paperfigs/paper_sharing.pdf")

mybins = 0:0.05:1
qnMTMT = calcquantile(dfMTMT, :npapers)
qndblp = calcquantile(dfdblp, :npapers)
close("all")
PyPlot.hist(qnMTMT, density=true, bins=mybins, label="MTMT", alpha=0.25)
PyPlot.hist(qndblp, density=true, bins=mybins, label="dblp", alpha=0.25)
xlabel("quantiles")
ylabel("density")
title("number of papers")
legend()
tight_layout()
savefig("../work/screen.pdf")

qgMTMT = calcquantile(dfMTMT, :groupprod);
qgdblp = calcquantile(dfdblp, :groupprod);
close("all")
PyPlot.hist(qgMTMT, density=true, bins=mybins, label="MTMT", alpha=0.25)
PyPlot.hist(qgdblp, density=true, bins=mybins, label="dblp", alpha=0.25)
xlabel("quantiles")
ylabel("density")
title("group productivity")
legend()
tight_layout()
savefig("../work/screen.pdf")

qwMTMT = calcquantile(dfMTMT, :wpapers);
qwdblp = calcquantile(dfdblp, :wpapers);
close("all")
PyPlot.hist(qwMTMT, density=true, bins=mybins, label="MTMT", alpha=0.25)
PyPlot.hist(qwdblp, density=true, bins=mybins, label="dblp", alpha=0.25)
xlabel("quantiles")
ylabel("density")
title("weighted number of papers")
legend()
tight_layout()
savefig("../work/screen.pdf")

# proportion of tight groups with significantly lower number of papers than
# random groups
sum(qnMTMT .<= 0.05) / length(qnMTMT)
sum(qndblp .<= 0.05) / length(qndblp)

# proportion of tight groups with significantly lower productivity than
# random groups
sum(qgMTMT .<= 0.05) / length(qgMTMT)
sum(qgdblp .<= 0.05) / length(qgdblp)

# proportion of tight groups with significantly lower weighted number of
# papers than random groups
sum(qwMTMT .<= 0.05) / length(qwMTMT)
sum(qwdblp .<= 0.05) / length(qwdblp)


